src/System.Management.Automation/security/CatalogHelper.cs

```diff
@@ -400,69 +398,62 @@ internal static string ProcessFilePathAttributeInCatalog(IntPtr memberAttrInfo)
         internal static string CalculateFileHash(string filePath, string hashAlgorithm)
         {
             string hashValue = string.Empty;
-            IntPtr catAdmin = IntPtr.Zero;
 
             // To get handle to the hash algorithm to be used to calculate hashes
-            if (!NativeMethods.CryptCATAdminAcquireContext2(ref catAdmin, IntPtr.Zero, hashAlgorithm, IntPtr.Zero, 0))
+            SafeCATAdminHandle catAdmin;
+            try
+            {
+                catAdmin = WinTrustMethods.CryptCATAdminAcquireContext2(hashAlgorithm);
+            }
+            catch (Win32Exception e)
             {
-                ErrorRecord errorRecord = new ErrorRecord(new InvalidOperationException(StringUtil.Format(CatalogStrings.UnableToAcquireHashAlgorithmContext, hashAlgorithm)), "UnableToAcquireHashAlgorithmContext", ErrorCategory.InvalidOperation, null);
+                ErrorRecord errorRecord = new ErrorRecord(
+                    new InvalidOperationException(StringUtil.Format(CatalogStrings.UnableToAcquireHashAlgorithmContext, hashAlgorithm), e),
+                    "UnableToAcquireHashAlgorithmContext",
+                    ErrorCategory.InvalidOperation,
+                    null);
                 _cmdlet.ThrowTerminatingError(errorRecord);
+                return hashValue;  // Satisfy using check below
             }
 
-            const DWORD GENERIC_READ = 0x80000000;
-            const DWORD OPEN_EXISTING = 3;
-            IntPtr INVALID_HANDLE_VALUE = new IntPtr(-1);
-
             // Open the file that is to be hashed for reading and get its handle
-            IntPtr fileHandle = NativeMethods.CreateFile(filePath, GENERIC_READ, 0, 0, OPEN_EXISTING, 0, IntPtr.Zero);
-            if (fileHandle != INVALID_HANDLE_VALUE)
+            FileStream fileStream;
+            try
+            {
+                fileStream = File.Open(filePath, FileMode.Open, FileAccess.Read);
+            }
+            catch (Exception e)
+            {
+                // If we are not able to open file that is to be hashed we can not continue with catalog validation
+                ErrorRecord errorRecord = new ErrorRecord(
+                    new InvalidOperationException(StringUtil.Format(CatalogStrings.UnableToReadFileToHash, filePath), e),
+                    "UnableToReadFileToHash",
+                    ErrorCategory.InvalidOperation,
+                    null);
+                _cmdlet.ThrowTerminatingError(errorRecord);
+                return hashValue;  // Satisfy using check below
+            }
+            
+            using (catAdmin)
+            using (fileStream)
```

Comment:
If we return in line 434 we don't dispose `catAdmin`. I'd suggest to refactor the method to make it more reliable.

Reply:
It ensures it gets disposed when no longer needed rather than wait for the garbage collector to get to it.